peer_strategies.go

```diff
@@ -20,40 +20,49 @@
 
 package tchannel
 
-import (
-	"math"
-	"math/rand"
-	"time"
-)
+import "math"
 
-type scoreCalculator interface {
+// ScoreCalculator defines the interface to calculate the score.
+type ScoreCalculator interface {
 	GetScore(p *Peer) uint64
 }
 
-type randCalculator struct {
-	rng *rand.Rand
-}
+// ScoreCalculatorFunc is an adapter that allows functions to be used as ScoreCalculator
+type ScoreCalculatorFunc func(p *Peer) uint64
 
-func (r *randCalculator) GetScore(p *Peer) uint64 {
-	return uint64(r.rng.Int63())
+// GetScore calls the underlying function.
+func (f ScoreCalculatorFunc) GetScore(p *Peer) uint64 {
+	return f(p)
 }
 
-func newRandCalculator() *randCalculator {
-	return &randCalculator{rng: NewRand(time.Now().UnixNano())}
+type zeroCalculator struct{}
+
+func (zeroCalculator) GetScore(p *Peer) uint64 {
+	return 0
 }
 
-type preferIncoming struct {
-	rng *rand.Rand
+func newZeroCalculator() zeroCalculator {
+	return zeroCalculator{}
 }
 
-func (r *preferIncoming) GetScore(p *Peer) uint64 {
-	rn := uint64(r.rng.Int31())
-	if p.NumInbound() > 0 {
-		rn += math.MaxUint32
+type preferIncomingCalculator struct{}
+
+func (preferIncomingCalculator) GetScore(p *Peer) uint64 {
+	if p.NumInbound() <= 0 {
+		return math.MaxUint64
+	}
+	count := 0
+	for _, c := range p.outboundConnections {
+		count = count + len(c.outbound.exchanges)
```

Comment:
accessing `c.outbound.exchanges` without read locks here and below. run with -race to see if these are detected